core/arch/mangle_shared.c

```diff
@@ -1016,11 +1016,15 @@ mangle(dcontext_t *dcontext, instrlist_t *ilist, uint *flags INOUT,
             mangle_possible_single_step(dcontext, ilist, instr);
             continue;
         }
-        else if (dcontext->single_step_addr != NULL &&
+        else if (dcontext->single_step_addr != NULL && instr_is_app(instr) &&
                  dcontext->single_step_addr == instr->translation) {
-            mangle_single_step(dcontext, ilist, *flags, instr);
-            /* Resets to generate single step exception only once. */
-            dcontext->single_step_addr = NULL;
+            instr_t * last_addr = instr_get_next_app(instr);
+            /* Checks if sandboxing added another app instruction. */
```

Comment:
I added this patch while testing with telock. There was a problem when there was sandboxing and single step on the same basic block.
Tell me if you want a separate pull request for this